src/meta/src/backup_restore/meta_snapshot_builder.rs

```diff
@@ -44,28 +45,40 @@ impl<S: MetaStore> MetaSnapshotBuilder<S> {
         }
     }
 
-    pub async fn build(&mut self, id: MetaSnapshotId) -> BackupResult<()> {
+    pub async fn build<D: Future<Output = BackupResult<HummockVersion>>>(
+        &mut self,
+        id: MetaSnapshotId,
+        hummock_version_checkpoint_builder: D,
+    ) -> BackupResult<()> {
         self.snapshot.format_version = VERSION;
         self.snapshot.id = id;
+        // Get `hummock_version_checkpoint` before `meta_store_snapshot`.
+        // We have ensure the required delta logs for replay is available, see
+        // `HummockManager::delete_version_deltas`.
+        let hummock_version_checkpoint = hummock_version_checkpoint_builder.await?;
         // Caveat: snapshot impl of etcd meta store doesn't prevent it from expiration.
         // So expired snapshot read may return error. If that happens,
         // tune auto-compaction-mode and auto-compaction-retention on demand.
         let meta_store_snapshot = self.meta_store.snapshot().await;
         let default_cf = self.build_default_cf(&meta_store_snapshot).await?;
         // hummock_version and version_stats is guaranteed to exist in a initialized cluster.
         let hummock_version = {
-            let mut redo_state = HummockVersion::list_at_snapshot::<S>(&meta_store_snapshot)
-                .await?
-                .into_iter()
-                .next()
-                .ok_or_else(|| anyhow!("hummock version checkpoint not found in meta store"))?;
+            let mut redo_state = hummock_version_checkpoint;
```

Comment:
```suggestion
            let mut redo_state = hummock_version_checkpoint_builder.await?;
```

Reply:
`hummock_version_checkpoint_builder.await?` is called before `self.meta_store.snapshot().await` by intention, to ensure resulted version is <= meta store snapshot, because subsequently we can always get a matching version for that meta store snapshot by replaying version delta.